bittide/src/Bittide/ClockControl/Callisto/Types.hs

```diff
@@ -0,0 +1,259 @@
+    -- ^ The controller remains in this state until stability has been
+    -- detected.
+  | Wait
+    -- ^ The controller remains in this state for the predefined
+    -- number of cycles with the assumption that the elastic buffers
+    -- of all other nodes are sufficiently stable after that time.
+      { targetCorrection :: !Float
+      -- ^ Stored correction value to be applied at reframing time.
+      , curWaitTime :: !(Unsigned 32)
+      -- ^ Number of cycles to wait until reframing takes place.
+      }
+  | Done
+    -- ^ Reframing has taken place. There is nothing more to do.
+  deriving (Generic, NFDataX)
+
+type instance SizeOf ReframingState =
+  SizeOf CUInt + SizeOf Float + SizeOf CUInt
+
+type instance Alignment ReframingState =
+  Alignment CUInt
+
+instance (SizeOf ReframingState ~ 12, Alignment ReframingState ~ 4)
+  => Storable ReframingState where
+  sizeOf = const $ natToNum @(SizeOf ReframingState)
+  alignment = const $ natToNum @(Alignment ReframingState)
+
+  peek p = (peekByteOff p 0 :: IO CUInt) >>= \case
+    0 -> return Detect
+    1 -> return Done
+    2 -> Wait <$> peekByteOff p 4
+              <*> ((fromIntegral :: CUInt -> Unsigned 32) <$> peekByteOff p 8)
+    _ -> error "out of range"
+
+  poke p = \case
+    Detect   -> pokeByteOff p 0 (0 :: CUInt)
+    Done     -> pokeByteOff p 0 (1 :: CUInt)
+    Wait{..} -> do
+      pokeByteOff p 0 (2 :: CUInt)
+      pokeByteOff p 4 targetCorrection
+      pokeByteOff p 8 (fromIntegral curWaitTime :: CUInt)
+
+-- | Callisto's internal state used in 'callisto'
+data ControlSt =
+  ControlSt
+    { _z_k :: !(Signed 32)
+    -- ^ Accumulated speed change requests, where speedup ~ 1, slowdown ~ -1.
+    , _b_k :: !SpeedChange
+    -- ^ Previously submitted speed change request. Used to determine the estimated
+    -- clock frequency.
+    , _steadyStateTarget :: !Float
+    -- ^ Steady-state value (determined when stability is detected for
+    -- the first time).
+    , rfState :: !ReframingState
+    -- ^ finite state machine for reframing detection
+    }
+  deriving (Generic, NFDataX)
+
+type instance SizeOf ControlSt =
+  SizeOf CInt + SizeOf SpeedChange + SizeOf Float + SizeOf ReframingState
+
+type instance Alignment ControlSt =
+  Alignment CInt
+
+instance (SizeOf ControlSt ~ 24, Alignment ControlSt ~ 4)
+  => Storable ControlSt where
+  sizeOf = const $ natToNum @(SizeOf ControlSt)
+  alignment = const $ natToNum @(Alignment ControlSt)
+
+  peek p =
+    ControlSt
+      <$> ((fromIntegral :: CInt -> Signed 32) <$> peekByteOff p 0)
+      <*> peekByteOff p 4
+      <*> peekByteOff p 8
+      <*> peekByteOff p 12
+
+  poke p ControlSt{..} = do
+    pokeByteOff p 0 (fromIntegral _z_k :: CInt)
+    pokeByteOff p 4 _b_k
+    pokeByteOff p 8 _steadyStateTarget
+    pokeByteOff p 12 rfState
+
+newtype VecS (n :: Nat) a = VecS (Vec n a)
+
+type instance SizeOf (VecS n a) =
+  SizeOf Int + n * SizeOf a
+
+type instance Alignment (VecS n a) =
+  Alignment a
```

Comment:
Hmm so this is tricky because there could be an `a` that has higher alignment than `Int`, so let's say 16! But later down there we first write the integer to the buffer and then all the elements at `n * sizeof a + sizeof int`, but that means we would end up with the elements starting at byte 4, would would be a wrong alignment!

Either we comment that it's only for types with alignment smaller than or equal to `Int` or we need to find a more generic version?

Reply:
We can add a constraint, which requires that the alignment of `a` is equals to that one of `Int`. Then we at least ensure correct usage at compile time.

Reply:
Yeah I think that would be good!